manage-data/ingest/transform-enrich/ingest-lag.md

```diff
@@ -0,0 +1,312 @@
+
+```json
+{
+  "event": {
+    "ingestion": {
+      "latency_logstash_to_elasticsearch": 443091,
+      "latency": 443093,
+      "latency_elastic_agent_to_logstash": 1
+    }
+  }
+}
+```
+
+#### Script
+
+If you want to remove the first calculation, you will need to ensure that the object `event.ingestion` is available.
+
+```json
+POST _ingest/pipeline/_simulate
+{
+    "docs": [{
+        "_source": {
+            "@timestamp": "2025-04-03T10:00:00.000Z",
+            "message": "user: philipp has logged in",
+            "_tmp": {
+              "logstash": "2025-04-03T10:00:02.456Z"
+            }
+
+        }
+    }],
+    "pipeline": {
+        "processors": [
+          {
+            "script": {
+              "description": "Calculates entire ingestion flow latency",
+              "if": "ctx['@timestamp'] != null",
+              "source": """
+                ZonedDateTime start = ZonedDateTime.parse(ctx['@timestamp']);
+                ctx.putIfAbsent("event", [:]);
+                ctx.event.putIfAbsent("ingestion", [:]);
+                ctx.event.ingestion.latency= ChronoUnit.SECONDS.between(start, metadata().now);
+              """
+            }
+          },
+          {
+            "script": {
+              "description": "Calculates logstash to Elasticsearch latency",
+              "if": "ctx._tmp?.logstash != null",
+              "source": """
+                ZonedDateTime start = ZonedDateTime.parse(ctx._tmp.logstash_seen);
+                ctx.event.ingestion.latency_logstash_to_elasticsearch=ChronoUnit.SECONDS.between(start, metadata().now);
+              """
+            }
+          },
+          {
+            "script": {
+              "description": "Calculates Elastic Agent to Logstash latency",
+              "if": "ctx._tmp?.logstash != null",
+              "source": """
+                ZonedDateTime start = ZonedDateTime.parse(ctx['@timestamp']);
+                ZonedDateTime end = ZonedDateTime.parse(ctx._tmp.logstash_seen);
+                ctx.event.ingestion.latency_elastic_agent_to_logstash=ChronoUnit.SECONDS.between(start, end);
+              """
+            }
+          }
+        ]
+    }
+}
+```
+
+### Elastic Agent => Logstash => Kafka => Logstash => Elasticsearch
+
+As with the previous scenario, adding an additional hop introduces another point where latency can occur. The recommendation here is to add another temporary timestamp field. For more details, refer to the [Elastic Agent => Logstash => Elasticsearch](#elastic-agent--logstash--elasticsearch) section above.
+
+Below is a script that calculates the latency for each step in the pipeline. The following values will be generated:
+
+```json
+{
+  "event": {
+    "ingestion": {
+      "latency_logstash_to_elasticsearch": 443091,
+      "latency_logstash_to_logstash": 1,
+      "latency": 443093,
+      "latency_elastic_agent_to_logstash": 1
+    }
+  }
+}
+```
+
+#### Script
```

Comment:
Ditto. Headings should be different to work in table of contents (anchors are generated automatically unless specified manually through square brackets notation.